graphql_compiler/interpreter/typedefs.py

```diff
@@ -0,0 +1,208 @@
+from abc import ABCMeta, abstractmethod
+from typing import (
+    AbstractSet,
+    Any,
+    Collection,
+    Generic,
+    Iterable,
+    Mapping,
+    Optional,
+    Tuple,
+    TypeVar,
+)
+
+from ..compiler.metadata import FilterInfo
+from ..typedefs import Literal
+
+
+DataToken = TypeVar("DataToken")
+
+
+class DataContext(Generic[DataToken]):
+    pass  # TODO(predrag): Will be implemented and documented in a later PR.
+
+
+EdgeDirection = Literal["in", "out"]
+EdgeInfo = Tuple[EdgeDirection, str]  # direction + edge name
+
+# TODO(predrag): Figure out a better type here. We need to balance between finding something
+#                easy and lightweight, and letting the user know about things like:
+#                optional edges, recursive edges, used fields/filters at the neighbor, etc.
+#                Will probably punt on this until the API is stabilized, since defining something
+#                here is not a breaking change.
+NeighborHint = Any
+
+
+class InterpreterAdapter(Generic[DataToken], metaclass=ABCMeta):
+    """Base class defining the API for schema-aware interpreter functionality over some schema.
+
+    This ABC is the abstraction through which the rest of the interpreter is schema-agnostic:
+    the rest of the interpreter code simply takes an instance of InterpreterAdapter and performs
+    all schema-aware operations through its simple, four-method API.
+
+    ## The DataToken type parameter
+
+    This class is generic on an implementer-chosen DataToken type, which to the rest of the library
+    represents an opaque reference to the data contained by a particular vertex in the data set
+    described by your chosen schema. For example, if building a subclass of InterpreterAdapter
+    called MyAdapter with dict as the DataToken type, MyAdapter should be defined as follows:
+
+        class MyAdapter(InterpreterAdapter[dict]):
+            ...
+
+    Here are a few common examples of DataToken types in practice:
+    - a dict containing the type name of the vertex and the values of all its properties;
+    - a dataclass containing the type name of the vertex, and the collection name and primary key
+      of the database entry using which the property values can be looked up, or
+    - an instance of a custom class which has *some* of the values of the vertex properties, and
+      has sufficient information to look up the rest of them if they are ever requested.
+
+    The best choice of DataToken type is dependent on the specific use case, e.g. whether the data
+    is already available in Python memory, or is on a local disk, or is a network hop away.
+
+    Implementers are free to choose any DataToken type and the interpreter code will happily use it.
+    However, for the sake of easier debugging and testing using the built-in functionality in
+    this library, it is desirable to make DataToken be a deep-copyable type that implements
+    equality beyond a simple referential equality check.
```

Comment:
I'd say "work best" instead of "only work", what do you think?

The debugging functionality makes deep copies because `DataToken` might be mutable, but in principle this can be avoided if the `DataToken` isn't deep-copyable. Deep-copying prevents referential equality from being that useful.

Debugging traces can always be read by a human, but if semantic equality isn't available then those traces might not be meaningfully comparable to each other — `==` might produce `False` even for semantically-identical traces, due to referential equality not matching.

Reply:
:+1: to "work best".

Reply:
(In retrospect, it probably would have been more correct from an "intelligible comment history" viewpoint to have created a new one--I'll try to remember this for next time.)